tests/versioning.go

```diff
@@ -1727,7 +1910,150 @@ func (s *VersioningIntegSuite) dispatchActivity(failMode activityFailMode, newVe
 		s.Equal("v1v2", out)
 	}
 
-	s.validateWorkflowBuildId(ctx, run.GetID(), run.GetRunID(), v1, newVersioning, true, "", nil)
+	s.validateWorkflowBuildId(ctx, run.GetID(), run.GetRunID(), v1, newVersioning, v1, "", nil)
+}
+
+func (s *VersioningIntegSuite) TestDispatchActivityUpgrade() {
+	// Testing:
+	// - redirecting dependent activity redirects wf
+	//   - this also implicitly verifies that sticky queue is cleared after redirecting activity task
+	// - redirecting independent activity does not redirect wf
+
+	tq := s.randomizeStr(s.T().Name())
+	v1 := s.prefixed("v1")
+	v11 := s.prefixed("v1.1")
+	v12 := s.prefixed("v1.2")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
+	defer cancel()
+
+	startedWf := make(chan struct{}, 1)
+	proceedWf := make(chan struct{}, 1)
+	started11 := make(chan struct{}, 1)
+	proceed11 := make(chan struct{}, 1)
+	started12 := make(chan struct{}, 1)
+	proceed12 := make(chan struct{}, 1)
+
+	act1 := func() (string, error) {
+		return "v1", nil
+	}
+	act11 := func() (string, error) {
+		started11 <- struct{}{}
+		s.waitForChan(ctx, proceed11)
+		return "v1.1", nil
+	}
+	act12 := func() (string, error) {
+		started12 <- struct{}{}
+		s.waitForChan(ctx, proceed12)
+		return "v1.2", nil
+	}
+	wf := func(ctx workflow.Context) (string, error) {
+		startedWf <- struct{}{}
+		// wait for adding v1 -> v1.1 redirect rule
+		<-proceedWf
+
+		var val1, val2 string
+		// run two activities
+		err1 := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			ScheduleToCloseTimeout: time.Minute,
+			DisableEagerExecution:  true,
+			VersioningIntent:       temporal.VersioningIntentCompatible,
+			StartToCloseTimeout:    1 * time.Second,
+		}), "act").Get(ctx, &val1)
+		s.NoError(err1)
+
+		startedWf <- struct{}{}
+		// wait for adding v1.1 -> v1.2 redirect rule
+		<-proceedWf
```

Comment:
should this wait on a signal to be safer? or is the point that you want to do it in between wft started and completed? if so maybe just comment why it's blocking at the go level instead of at the wf level

Reply:
yeah, that's the case. Updated the comment.